### PR TITLE
[Feature] Use new data format of memtable for multi-version

### PR DESCRIPTION
**Pull request by TszKitLo40**
## Why I'm doing:
In #33441, memtable has been implemented. But in this implementation, we need to encode the version in the key. When we want to read the latest version, we may need to scan all the map and decode every key which may cause poor performance. The most common scenarios in pk index is to get the latest version now.
## What I'm doing:
We design a new data format for memtable. We use `std::list<pair<int64_t, IndexValue>>` for multi-versions. The latest version will always the front of the list so we do not need to decode the key. To support multi-version in future, the previous versions will also be stored in the list. We only support storing the latest version now. Version is passed by `LakePersistentIndex`.

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: be/src/storage/lake/persistent_index_memtable.cpp [hunk 1]
```diff
@@ -16,82 +16,110 @@
 
 namespace starrocks::lake {
 
+void PersistentIndexMemtable::insert(const std::string& key, int64_t version, const IndexValue& value) {
+    std::list<IndexValueInfo> index_value_infos;
+    index_value_infos.emplace_front(version, value);
+    _map.emplace(key, index_value_infos);
+}
+
+void PersistentIndexMemtable::update(std::list<IndexValueInfo>& index_value_infos, int64_t version,
```
Comment (sduzh) [inline be/src/storage/lake/persistent_index_memtable.cpp:25]:
```suggestion
void PersistentIndexMemtable::update(std::list<IndexValueInfo>* index_value_infos, int64_t version,
```
Normally we don't pass non-const reference in StarRocks.

File: be/src/storage/lake/persistent_index_memtable.cpp [hunk 2]
```diff
@@ -16,82 +16,110 @@
...
+                                     const IndexValue& value) {
+    std::list<IndexValueInfo> t;
+    t.emplace_front(version, value);
+    index_value_infos.swap(t);
+}
+
 Status PersistentIndexMemtable::upsert(size_t n, const Slice* keys, const IndexValue* values, IndexValue* old_values,
-                                       KeyIndexesInfo* not_found, size_t* num_found) {
+                                       KeyIndexesInfo* not_found, size_t* num_found, int64_t version) {
```
Comment (sduzh) [inline be/src/storage/lake/persistent_index_memtable.cpp:33]: How about defining `keys` as type  `std::span<const Slice>`?

File: be/src/storage/lake/persistent_index_memtable.cpp [hunk 3]
```diff
@@ -16,82 +16,110 @@
...
-            it->second = value;
+            update(it->second, version, value);
         }
     }
     *num_found = nfound;
     return Status::OK();
 }
 
-Status PersistentIndexMemtable::insert(size_t n, const Slice* keys, const IndexValue* values) {
+Status PersistentIndexMemtable::insert(size_t n, const Slice* keys, const IndexValue* values, int64_t version) {
```
Comment (sduzh) [inline be/src/storage/lake/persistent_index_memtable.cpp:54]: Nit: `std::span<const Slice> keys`

File: be/src/storage/lake/persistent_index_memtable.h [hunk 4]
```diff
@@ -14,30 +14,44 @@
...
 namespace starrocks::lake {
 
+using IndexValueInfo = std::pair<int64_t, IndexValue>;
+
 class PersistentIndexMemtable {
 public:
+    // |version|: version of index values, -1 means latest version
     Status upsert(size_t n, const Slice* keys, const IndexValue* values, IndexValue* old_values,
-                  KeyIndexesInfo* not_found, size_t* num_found);
+                  KeyIndexesInfo* not_found, size_t* num_found, int64_t version = -1);
```
Comment (sduzh) [inline be/src/storage/lake/persistent_index_memtable.h:29]: replace `-1` with named constant

File: be/src/storage/lake/persistent_index_memtable.cpp (hunk 2 above)
Comment (TszKitLo40) [reply be/src/storage/lake/persistent_index_memtable.cpp:33]: It seems that const Slice* can not be casted to std::span<const Slice>

File: be/src/storage/lake/persistent_index_memtable.cpp (hunk 1 above)
Comment (TszKitLo40) [reply be/src/storage/lake/persistent_index_memtable.cpp]: done

**Review by sduzh — DISMISSED** [review]

File: be/src/storage/lake/persistent_index_memtable.h [hunk 5]
```diff
@@ -14,30 +14,44 @@
...
-    Status get(size_t n, const Slice* keys, IndexValue* values, KeyIndexesInfo* not_found, size_t* num_found);
+    // |version|: version of index values
+    Status get(size_t n, const Slice* keys, IndexValue* values, KeyIndexesInfo* not_found, size_t* num_found,
+               int64_t version);
 
     void clear();
 
 private:
-    std::map<std::string, IndexValue, std::less<>> _map;
+    void insert(const std::string& key, int64_t version, const IndexValue& value);
```
Comment (luohaha) [inline be/src/storage/lake/persistent_index_memtable.h:49]: Use `_ insert` and `_upsert` to distinguish private and publish func？

File: be/src/storage/lake/persistent_index_memtable.cpp [hunk 6]
```diff
@@ -16,31 +16,43 @@
...
+        std::list<IndexValueInfo> index_value_infos;
+        index_value_infos.emplace_front(version, value);
+        if (auto [it, inserted] = _map.emplace(key, index_value_infos); inserted) {
             not_found->key_index_infos.emplace_back(i);
         } else {
-            auto old_value = it->second;
+            auto& old_index_value_infos = it->second;
+            auto old_value = old_index_value_infos.front().second;
             old_values[i] = old_value;
             nfound += old_value.get_value() != NullIndexValue;
```
Comment (luohaha) [inline be/src/storage/lake/persistent_index_memtable.cpp:40]: execute `not_found->key_index_infos.emplace_back(i);` when `get_value() == NullIndexValue` ?

**Review by luohaha — APPROVED** [review]

**Review by sduzh — APPROVED** [review]

**Pull request merged**